File: pdftransl/quality/reviewer.py
"""LLM-ревью — вторая линия самоконтроля.

Ревьюер перепроверяет проблемные сегменты: JSON-вердикт
{ok: true} либо {ok: false, revised: ...}. Ревизия принимается
только если не теряет содержимое плейсхолдеров (формулы/ссылки),
иначе остаётся исходный перевод с пометкой. Сбой ревью не фатален —
пайплайн продолжает без него.
"""

from __future__ import annotations

import json
import logging
import re
from typing import Optional

from pdftransl.config import PipelineConfig
from pdftransl.llm.base import BaseLLMClient
from pdftransl.masking import unmask
from pdftransl.models import QAIssue, Segment
from pdftransl.quality.validators import validate_segment
from pdftransl.translation.prompts import REVIEW_SYSTEM, REVIEW_USER, lang_name

logger = logging.getLogger(__name__)

_JSON_RE = re.compile(r"\{.*\}", re.DOTALL)


class Reviewer:
    def __init__(self, client: BaseLLMClient, config: PipelineConfig):
        self.client = client
        self.config = config

    def review_segment(self, segment: Segment) -> Segment:
        """Ask the LLM to critique/fix a translated segment in place."""
        if segment.kind != "translate" or not segment.translation:
            return segment
        system = REVIEW_SYSTEM.format(
            src=lang_name(self.config.source_lang),
            tgt=lang_name(self.config.target_lang),
        )
        user = REVIEW_USER.format(
            source=segment.masked_text or segment.source_text,
            translation=segment.translation,
        )
        response_format = (
            {"type": "json_object"} if self.config.structured_outputs else None
        )
        try:
            raw = self.client.chat(
                [
                    {"role": "system", "content": system},
                    {"role": "user", "content": user},
                ],
                temperature=0.0,
                response_format=response_format,
            )
        except Exception as exc:  # review must never break the pipeline
            logger.warning("Review failed for %s: %s", segment.id, exc)
            segment.issues.append(
                QAIssue("review_error", f"review pass failed: {exc}", "warning")
            )
            return segment

        verdict = _parse_review(raw)
        if verdict is None:
            segment.issues.append(
                QAIssue("review_unparsed", "reviewer returned unparsable output", "warning")
            )
            return segment

        if verdict.get("ok"):
            segment.issues.append(QAIssue("reviewed_ok", "approved by LLM reviewer", "info"))
            return segment

        revised = verdict.get("revised")
        # модель может вернуть revised не-строкой (null, объект) — это
        # «ревизии нет», а не повод падать
        if revised and isinstance(revised, str):
            restored, missing, unknown = unmask(revised.strip(), segment.placeholders)
            # The reviewer sees the *restored* translation (real formulas,
            # not tokens), so a good revision usually carries the actual
            # content instead of ⟦PH…⟧ tokens. A token is only truly
            # missing if its content is absent too — otherwise every
            # revision of a placeholder-bearing segment would be rejected.
            missing = [
                t for t in missing if segment.placeholders[t] not in restored
            ]
            # accept the revision only if it doesn't break placeholders
            if not missing and not unknown:
                segment.translation = restored
                segment.issues = validate_segment(segment, self.config)
                # (verdict.get("notes") or ""): при {"notes": null} .get
                # возвращает None (ключ есть!) и срез падал с TypeError,
                # обрывая ревью всего документа
                notes = verdict.get("notes")
                if not isinstance(notes, str):
                    notes = ""
                segment.issues.append(QAIssue(
                    "reviewed_revised",
                    f"revised by LLM reviewer: {notes[:300]}",
                    "info",
                ))
            else:
                segment.issues.append(QAIssue(
                    "review_rejected",
                    "reviewer revision dropped placeholders; kept original",
                    "warning",
                ))
        return segment

    def review_segments(
        self, segments: list[Segment], only_flagged: bool = True
    ) -> list[Segment]:
        for segment in segments:
            if segment.kind != "translate":
                continue
            has_flags = any(i.severity in ("warning", "error") for i in segment.issues)
            if only_flagged and not has_flags:
                continue
            # Изоляция: неожиданное исключение на одном сегменте (кривой
            # JSON модели и т.п.) раньше обрывало ревью ВСЕХ оставшихся
            # сегментов — стадия падала целиком.
            try:
                self.review_segment(segment)
            except Exception as exc:  # noqa: BLE001 - ревью не критично
                logger.warning("Review crashed on segment %s: %s", segment.id, exc)
                segment.issues.append(
                    QAIssue("review_error", f"review pass crashed: {exc}", "warning")
                )
        return segments


def _parse_review(raw: str) -> Optional[dict]:
    raw = raw.strip()
    if raw.startswith("```"):
        raw = re.sub(r"^```[a-z]*\n?|\n?```$", "", raw)

    def _as_dict(value) -> Optional[dict]:
        # json.loads может вернуть null/строку/список — для вердикта
        # годится только объект, остальное = «не распарсили»
        return value if isinstance(value, dict) else None

    try:
        return _as_dict(json.loads(raw))
    except ValueError:
        match = _JSON_RE.search(raw)
        if match:
            try:
                return _as_dict(json.loads(match.group(0)))
            except ValueError:
                return None
    return None

File: cpp/src/quality/reviewer.cpp
#include "quality/reviewer.h"
#include "core/config.h"
#include "quality/validators.h"
#include "translation/prompts.h"
#include <QJsonDocument>
#include <QJsonObject>
#include <QJsonValue>
#include <QRegularExpression>
#include <QSet>
#include <algorithm>
#include <optional>

namespace pdftransl {

namespace {

QSet<QString> placeholderTokensIn(const QString& text) {
    QSet<QString> tokens;
    static const QRegularExpression re(QStringLiteral("⟦PH\\d+⟧"));
    auto it = re.globalMatch(text);
    while (it.hasNext()) tokens.insert(it.next().captured(0));
    return tokens;
}

QString stripFence(const QString& raw) {
    QString text = raw.trimmed();
    if (text.startsWith("```")) {
        static const QRegularExpression fenceStart(QStringLiteral("^```[a-zA-Z]*\\n?"));
        static const QRegularExpression fenceEnd(QStringLiteral("\\n?```$"));
        text.remove(fenceStart);
        text.remove(fenceEnd);
    }
    return text;
}

// Модель могла обернуть JSON пояснением или markdown-оградой — сначала
// пробуем распарсить весь ответ, затем вытаскиваем первый похожий на объект
// фрагмент.
std::optional<QJsonObject> parseVerdict(const QString& raw) {
    const QString text = stripFence(raw);
    QJsonParseError err{};
    QJsonDocument doc = QJsonDocument::fromJson(text.toUtf8(), &err);
    if (err.error == QJsonParseError::NoError && doc.isObject()) return doc.object();

    static const QRegularExpression jsonRe(QStringLiteral("\\{[\\s\\S]*\\}"));
    auto match = jsonRe.match(text);
    if (match.hasMatch()) {
        QJsonDocument doc2 = QJsonDocument::fromJson(match.captured(0).toUtf8(), &err);
        if (err.error == QJsonParseError::NoError && doc2.isObject()) return doc2.object();
    }
    return std::nullopt;
}

} // namespace

Reviewer::Reviewer(LLMClientPtr client) : m_client(std::move(client)) {}

Segment Reviewer::reviewOne(Segment segment, const QString& sourceLang, const QString& targetLang) {
    if (segment.passthrough || segment.translation.trimmed().isEmpty()) return segment;

    const QString system = REVIEW_SYSTEM.arg(langName(sourceLang), langName(targetLang));
    const QString sourceForReview =
        segment.maskedText.isEmpty() ? segment.sourceText : segment.maskedText;
    const QString user = REVIEW_USER.arg(sourceForReview, segment.translation);

    QString raw;
    try {
        raw = m_client->chat({Message{"system", system}, Message{"user", user}}, 0.0);
    } catch (const std::exception& exc) {
        segment.issues.push_back(QAIssue{
            "warning", QStringLiteral("review pass failed: %1").arg(exc.what()), "review_error"});
        return segment;
    }

    auto verdict = parseVerdict(raw);
    if (!verdict) {
        segment.issues.push_back(
            QAIssue{"warning", "reviewer returned unparsable output", "review_unparsed"});
        return segment;
    }

    if (verdict->value("ok").toBool(false)) {
        segment.issues.push_back(QAIssue{"info", "approved by LLM reviewer", "reviewed_ok"});
        return segment;
    }

    const QJsonValue revisedVal = verdict->value("revised");
    if (revisedVal.isString() && !revisedVal.toString().trimmed().isEmpty()) {
        const QString revised = revisedVal.toString().trimmed();

        // Ревизия принимается только если не теряет ни одного плейсхолдера,
        // присутствовавшего в замаскированном источнике сегмента (формулы,
        // ссылки, код).
        const QSet<QString> required = placeholderTokensIn(sourceForReview);
        const QSet<QString> present = placeholderTokensIn(revised);
        bool lost = false;
        for (const auto& token : required) {
            if (!present.contains(token)) { lost = true; break; }
        }

        if (!lost) {
            segment.translation = revised;
            PipelineConfig cfg; // только языковая пара важна для валидатора
            cfg.sourceLang = sourceLang;
            cfg.targetLang = targetLang;
            segment.issues = validateSegment(segment, cfg);
            segment.ok = std::none_of(segment.issues.begin(), segment.issues.end(),
                                       [](const QAIssue& i) { return i.level == "error"; });
            QString notes = verdict->value("notes").toString();
            segment.issues.push_back(QAIssue{
                "info",
                QStringLiteral("revised by LLM reviewer: %1").arg(notes.left(300)),
                "reviewed_revised"});
        } else {
            segment.issues.push_back(QAIssue{
                "warning", "reviewer revision dropped placeholders; kept original",
                "review_rejected"});
        }
    }
    return segment;
}

std::vector<Segment> Reviewer::reviewSegments(const std::vector<Segment>& flagged,
                                               const QString& sourceLang,
                                               const QString& targetLang) {
    std::vector<Segment> result;
    result.reserve(flagged.size());
    for (const auto& segment : flagged) {
        // Изоляция: сбой на одном сегменте (кривой JSON модели и т.п.) не
        // должен обрывать ревью остальных.
        try {
            result.push_back(reviewOne(segment, sourceLang, targetLang));
        } catch (const std::exception& exc) {
            Segment copy = segment;
            copy.issues.push_back(QAIssue{
                "warning", QStringLiteral("review pass crashed: %1").arg(exc.what()), "review_error"});
            result.push_back(copy);
        }
    }
    return result;
}

} // namespace pdftransl
